src/tests/tribol_enzyme_nodal_normal.cpp

```diff
@@ -0,0 +1,145 @@
+    norm.Compute( mesh );
+    auto num_dofs = mesh.numberOfNodes() * mesh.spatialDimension();
+    mfem::SparseMatrix dndx( num_dofs );
+    auto& dndx_data = norm.getJacobianData();
+    // get nonmortar/nonmortar contributions
+    auto& elem_Js =
+        dndx_data.getBlockJ()( static_cast<int>( BlockSpace::NONMORTAR ), static_cast<int>( BlockSpace::NONMORTAR ) );
+    // assemble element contributions to a global sparse matrix
+    auto num_nodes_per_el = mesh.numberOfNodesPerElement();
+    for ( int i{ 0 }; i < mesh.numberOfElements(); ++i ) {
+      for ( int d1{ 0 }; d1 < mesh.spatialDimension(); ++d1 ) {
+        for ( int j{ 0 }; j < num_nodes_per_el; ++j ) {
+          auto global_idx_j = mesh.numberOfNodes() * d1 + mesh.getGlobalNodeId( i, j );
+          auto local_idx_j = num_nodes_per_el * d1 + j;
+          for ( int d2{ 0 }; d2 < mesh.spatialDimension(); ++d2 ) {
+            for ( int k{ 0 }; k < num_nodes_per_el; ++k ) {
+              auto global_idx_k = mesh.numberOfNodes() * d2 + mesh.getGlobalNodeId( i, k );
+              auto local_idx_k = num_nodes_per_el * d2 + k;
+              dndx.Add( global_idx_j, global_idx_k, elem_Js[i]( local_idx_j, local_idx_k ) );
+            }
+          }
+        }
+      }
+    }
+    dndx.Finalize();
+    mfem::DenseMatrix dndx_enzyme;
+    dndx.ToDenseMatrix( dndx_enzyme );
+
+    mfem::DenseMatrix dndx_fd( num_dofs );
+    auto mesh_view = mesh.getView();
+    norm = VertexAvgNormal( false );
+    Array2D<RealT> n_base = mesh_view.getNodalNormals();
+    for ( int dx{ 0 }; dx < mesh.spatialDimension(); ++dx ) {
+      for ( int nx{ 0 }; nx < mesh.numberOfNodes(); ++nx ) {
+        auto x_idx = dx * mesh.numberOfNodes() + nx;
+        x[dx * mesh.numberOfNodes() + nx] += delta_;
+        norm.Compute( mesh );
+        auto local_mesh_view = mesh.getView();
+        for ( int dn{ 0 }; dn < mesh.spatialDimension(); ++dn ) {
+          for ( int nn{ 0 }; nn < mesh.numberOfNodes(); ++nn ) {
+            auto n_idx = dn * mesh.numberOfNodes() + nn;
+            dndx_fd( n_idx, x_idx ) = ( local_mesh_view.getNodalNormals()( dn, nn ) - n_base( dn, nn ) ) / delta_;
+          }
+        }
+        x[dx * mesh.numberOfNodes() + nx] -= delta_;
+      }
+    }
+
+    // write deltas to screen
+    std::cout << "dn/dx ------------------------------" << std::endl;
+    for ( int i{ 0 }; i < num_dofs; ++i ) {
+      for ( int j{ 0 }; j < num_dofs; ++j ) {
+        auto diff = std::abs( dndx_enzyme( i, j ) - dndx_fd( i, j ) );
+        if ( diff > delta_ ) {
+          std::cout << "  (" << i << ", " << j << ") : Diff: " << diff
+                    << "   Ratio: " << dndx_enzyme( i, j ) / dndx_fd( i, j ) << "   Enzyme: " << dndx_enzyme( i, j )
+                    << "   FD: " << dndx_fd( i, j ) << std::endl;
+        }
+        EXPECT_NEAR( dndx_enzyme( i, j ), dndx_fd( i, j ), delta_ );
+      }
+    }
+  }
+};
+
+TEST_F( EnzymeNodalNormalTest, TwoElementsFlatNormalJacobian )
```

Comment:
Is it worth testing a warped element configuration?

Reply:
We test warped elements with the complete Jacobian, so it doesn't seem necessary for now.  We can add more tests later if needed.